File: sol-trade-sdk-python/src/instruction/raydium_amm_v4_builder.py
"""
Raydium AMM V4 instruction builder for Solana trading SDK.
Production-grade implementation with all constants, discriminators, and PDA derivation functions.
"""

from typing import List, Optional
from dataclasses import dataclass
from solders.pubkey import Pubkey
from solders.instruction import Instruction, AccountMeta
import struct

from .common import (
    TOKEN_PROGRAM,
    WSOL_TOKEN_ACCOUNT,
    USDC_TOKEN_ACCOUNT,
    DEFAULT_SLIPPAGE,
    get_associated_token_address,
    create_associated_token_account_idempotent_instruction,
    handle_wsol,
    close_wsol,
    close_token_account_instruction,
    calculate_with_slippage_sell,
)

# ============================================
# Raydium AMM V4 Program ID
# ============================================

RAYDIUM_AMM_V4_PROGRAM_ID: Pubkey = Pubkey.from_string("675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8")

# ============================================
# Raydium AMM V4 Constants
# ============================================

# Authority
AUTHORITY: Pubkey = Pubkey.from_string("5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1")

# Fee Rates
TRADE_FEE_NUMERATOR: int = 25
TRADE_FEE_DENOMINATOR: int = 10000
SWAP_FEE_NUMERATOR: int = 25
SWAP_FEE_DENOMINATOR: int = 10000

# ============================================
# Instruction Discriminators
# ============================================

# Note: Raydium AMM V4 uses single-byte discriminators
SWAP_BASE_IN_DISCRIMINATOR: bytes = bytes([9])
SWAP_BASE_OUT_DISCRIMINATOR: bytes = bytes([11])

# ============================================
# Seeds
# ============================================

POOL_SEED = b"pool"


# ============================================
# Raydium AMM V4 Parameters Dataclass
# ============================================

@dataclass
class RaydiumAmmV4Params:
    """Parameters for Raydium AMM V4 protocol trading."""
    amm: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    coin_mint: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    pc_mint: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    token_coin: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    token_pc: Pubkey = Pubkey.from_string("11111111111111111111111111111111")
    coin_reserve: int = 0
    pc_reserve: int = 0

    @property
    def is_wsol(self) -> bool:
        """Check if the pool contains WSOL."""
        return self.coin_mint == WSOL_TOKEN_ACCOUNT or self.pc_mint == WSOL_TOKEN_ACCOUNT

    @property
    def is_usdc(self) -> bool:
        """Check if the pool contains USDC."""
        return self.coin_mint == USDC_TOKEN_ACCOUNT or self.pc_mint == USDC_TOKEN_ACCOUNT


# ============================================
# Raydium AMM V4 Calculation Functions
# ============================================

def compute_swap_amount(
    coin_reserve: int,
    pc_reserve: int,
    is_coin_in: bool,
    amount_in: int,
    slippage_bps: int,
) -> tuple:
    """
    Compute swap output amount for Raydium AMM V4.
    Returns (amount_out, min_amount_out).
    """
    # Apply trade fee (0.25%)
    amount_in_after_fee = amount_in - (amount_in * TRADE_FEE_NUMERATOR) // TRADE_FEE_DENOMINATOR

    if is_coin_in:
        # Coin -> PC
        # out = (amount_in * pc_reserve) / (coin_reserve + amount_in)
        numerator = amount_in_after_fee * pc_reserve
        denominator = coin_reserve + amount_in_after_fee
        amount_out = numerator // denominator
    else:
        # PC -> Coin
        # out = (amount_in * coin_reserve) / (pc_reserve + amount_in)
        numerator = amount_in_after_fee * coin_reserve
        denominator = pc_reserve + amount_in_after_fee
        amount_out = numerator // denominator

    # Apply slippage
    min_amount_out = calculate_with_slippage_sell(amount_out, slippage_bps)

    return (amount_out, min_amount_out)


# ============================================
# Build Buy Instructions
# ============================================

def build_buy_instructions(
    payer: Pubkey,
    output_mint: Pubkey,
    input_amount: int,
    params: RaydiumAmmV4Params,
    slippage_bps: int = DEFAULT_SLIPPAGE,
    create_input_ata: bool = True,
    create_output_ata: bool = True,
    close_input_ata: bool = False,
    fixed_output_amount: Optional[int] = None,
) -> List[Instruction]:
    """
    Build Raydium AMM V4 buy instructions.

    Args:
        payer: The wallet paying for the swap
        output_mint: The token mint to buy
        input_amount: Amount of SOL/USDC to spend
        params: Raydium AMM V4 protocol parameters
        slippage_bps: Slippage tolerance in basis points
        create_input_ata: Whether to create WSOL ATA if needed
        create_output_ata: Whether to create output token ATA if needed
        close_input_ata: Whether to close WSOL ATA after swap
        fixed_output_amount: If set, use this as exact output amount

    Returns:
        List of instructions for the buy operation
    """
    if input_amount == 0:
        raise ValueError("Amount cannot be zero")

    instructions = []

    # Validate pool contains WSOL or USDC
    if not params.is_wsol and not params.is_usdc:
        raise ValueError("Pool must contain WSOL or USDC")

    # Determine if coin is input (WSOL/USDC)
    is_coin_in = params.coin_mint == WSOL_TOKEN_ACCOUNT or params.coin_mint == USDC_TOKEN_ACCOUNT

    # Calculate swap amount
    _, min_amount_out = compute_swap_amount(
        params.coin_reserve,
        params.pc_reserve,
        is_coin_in,
        input_amount,
        slippage_bps,
    )

    if fixed_output_amount is not None:
        minimum_amount_out = fixed_output_amount
    else:
        minimum_amount_out = min_amount_out

    # Determine input mint (WSOL or USDC)
    input_mint = WSOL_TOKEN_ACCOUNT if params.is_wsol else USDC_TOKEN_ACCOUNT

    # Get user token accounts
    user_source_token_account = get_associated_token_address(payer, input_mint, TOKEN_PROGRAM)
    user_destination_token_account = get_associated_token_address(payer, output_mint, TOKEN_PROGRAM)

    # Handle WSOL if needed
    if create_input_ata and params.is_wsol:
        instructions.extend(handle_wsol(payer, input_amount))

    # Create output ATA if needed
    if create_output_ata:
        instructions.append(
            create_associated_token_account_idempotent_instruction(
                payer, payer, output_mint, TOKEN_PROGRAM
            )
        )

    # Build instruction data (1 byte discriminator + 8 bytes amount_in + 8 bytes min_out)
    data = SWAP_BASE_IN_DISCRIMINATOR + struct.pack("<QQ", input_amount, minimum_amount_out)

    # Build accounts list (17 accounts)
    # Note: Raydium AMM V4 has specific account ordering
    accounts = [
        AccountMeta(TOKEN_PROGRAM, False, False),  # token_program (readonly)
        AccountMeta(params.amm, False, True),  # amm (writable)
        AccountMeta(AUTHORITY, False, False),  # authority (readonly)
        AccountMeta(params.amm, False, False),  # amm_open_orders (uses amm address)
        AccountMeta(params.token_coin, False, True),  # pool_coin_token_account (writable)
        AccountMeta(params.token_pc, False, True),  # pool_pc_token_account (writable)
        AccountMeta(params.amm, False, False),  # serum_program (placeholder)
        AccountMeta(params.amm, False, False),  # serum_market (placeholder)
        AccountMeta(params.amm, False, False),  # serum_bids (placeholder)
        AccountMeta(params.amm, False, False),  # serum_asks (placeholder)
        AccountMeta(params.amm, False, False),  # serum_event_queue (placeholder)
        AccountMeta(params.amm, False, False),  # serum_coin_vault_account (placeholder)
        AccountMeta(params.amm, False, False),  # serum_pc_vault_account (placeholder)
        AccountMeta(params.amm, False, False),  # serum_vault_signer (placeholder)
        AccountMeta(user_source_token_account, False, True),  # user_source_token_account (writable)
        AccountMeta(user_destination_token_account, False, True),  # user_destination_token_account (writable)
        AccountMeta(payer, True, False),  # user_source_owner (signer)
    ]

    instructions.append(Instruction(RAYDIUM_AMM_V4_PROGRAM_ID, data, accounts))

    # Close WSOL ATA if requested
    if close_input_ata and params.is_wsol:
        instructions.extend(close_wsol(payer))

    return instructions


# ============================================
# Build Sell Instructions
# ============================================

def build_sell_instructions(
    payer: Pubkey,
    input_mint: Pubkey,
    input_amount: int,
    params: RaydiumAmmV4Params,
    slippage_bps: int = DEFAULT_SLIPPAGE,
    create_output_ata: bool = True,
    close_output_ata: bool = False,
    close_input_ata: bool = False,
    fixed_output_amount: Optional[int] = None,
) -> List[Instruction]:
    """
    Build Raydium AMM V4 sell instructions.

    Args:
        payer: The wallet paying for the swap
        input_mint: The token mint to sell
        input_amount: Amount of tokens to sell
        params: Raydium AMM V4 protocol parameters
        slippage_bps: Slippage tolerance in basis points
        create_output_ata: Whether to create WSOL ATA for receiving SOL
        close_output_ata: Whether to close WSOL ATA after swap
        close_input_ata: Whether to close token ATA after swap
        fixed_output_amount: If set, use this as exact output amount

    Returns:
        List of instructions for the sell operation
    """
    if input_amount == 0:
        raise ValueError("Amount cannot be zero")

    instructions = []

    # Validate pool contains WSOL or USDC
    if not params.is_wsol and not params.is_usdc:
        raise ValueError("Pool must contain WSOL or USDC")

    # Determine if pc is output (WSOL/USDC)
    # For sell: is_base_in = True means we're selling PC to get Coin
    # is_base_in = False means we're selling Coin to get PC
    is_pc_out = params.pc_mint == WSOL_TOKEN_ACCOUNT or params.pc_mint == USDC_TOKEN_ACCOUNT

    # Calculate swap amount (is_pc_out reversed because we're selling to get WSOL/USDC)
    _, min_amount_out = compute_swap_amount(
        params.coin_reserve,
        params.pc_reserve,
        not is_pc_out,  # Reversed for sell
        input_amount,
        slippage_bps,
    )

    if fixed_output_amount is not None:
        minimum_amount_out = fixed_output_amount
    else:
        minimum_amount_out = min_amount_out

    # Determine output mint (WSOL or USDC)
    output_mint = WSOL_TOKEN_ACCOUNT if params.is_wsol else USDC_TOKEN_ACCOUNT

    # Get user token accounts
    user_source_token_account = get_associated_token_address(payer, input_mint, TOKEN_PROGRAM)
    user_destination_token_account = get_associated_token_address(payer, output_mint, TOKEN_PROGRAM)

    # Create WSOL ATA if needed for receiving SOL
    if create_output_ata and params.is_wsol:
        instructions.append(
            create_associated_token_account_idempotent_instruction(
                payer, payer, WSOL_TOKEN_ACCOUNT, TOKEN_PROGRAM
            )
        )

    # Build instruction data (1 byte discriminator + 8 bytes amount_in + 8 bytes min_out)
    data = SWAP_BASE_IN_DISCRIMINATOR + struct.pack("<QQ", input_amount, minimum_amount_out)

    # Build accounts list (17 accounts)
    accounts = [
        AccountMeta(TOKEN_PROGRAM, False, False),  # token_program (readonly)
        AccountMeta(params.amm, False, True),  # amm (writable)
        AccountMeta(AUTHORITY, False, False),  # authority (readonly)
        AccountMeta(params.amm, False, False),  # amm_open_orders (uses amm address)
        AccountMeta(params.token_coin, False, True),  # pool_coin_token_account (writable)
        AccountMeta(params.token_pc, False, True),  # pool_pc_token_account (writable)
        AccountMeta(params.amm, False, False),  # serum_program (placeholder)
        AccountMeta(params.amm, False, False),  # serum_market (placeholder)
        AccountMeta(params.amm, False, False),  # serum_bids (placeholder)
        AccountMeta(params.amm, False, False),  # serum_asks (placeholder)
        AccountMeta(params.amm, False, False),  # serum_event_queue (placeholder)
        AccountMeta(params.amm, False, False),  # serum_coin_vault_account (placeholder)
        AccountMeta(params.amm, False, False),  # serum_pc_vault_account (placeholder)
        AccountMeta(params.amm, False, False),  # serum_vault_signer (placeholder)
        AccountMeta(user_source_token_account, False, True),  # user_source_token_account (writable)
        AccountMeta(user_destination_token_account, False, True),  # user_destination_token_account (writable)
        AccountMeta(payer, True, False),  # user_source_owner (signer)
    ]

    instructions.append(Instruction(RAYDIUM_AMM_V4_PROGRAM_ID, data, accounts))

    # Close WSOL ATA if requested
    if close_output_ata and params.is_wsol:
        instructions.extend(close_wsol(payer))

    # Close token ATA if requested
    if close_input_ata:
        instructions.append(
            close_token_account_instruction(
                TOKEN_PROGRAM,
                user_source_token_account,
                payer,
                payer,
            )
        )

    return instructions


# ============================================
# Exports
# ============================================

__all__ = [
    # Program IDs and Constants
    "RAYDIUM_AMM_V4_PROGRAM_ID",
    "AUTHORITY",
    "TRADE_FEE_NUMERATOR",
    "TRADE_FEE_DENOMINATOR",
    "SWAP_FEE_NUMERATOR",
    "SWAP_FEE_DENOMINATOR",
    # Discriminators
    "SWAP_BASE_IN_DISCRIMINATOR",
    "SWAP_BASE_OUT_DISCRIMINATOR",
    # Params
    "RaydiumAmmV4Params",
    # Calculation Functions
    "compute_swap_amount",
    # Instruction Builders
    "build_buy_instructions",
    "build_sell_instructions",
]

File: sol-trade-sdk-nodejs/src/instruction/raydium_amm_v4_builder.ts
/**
 * Raydium AMM V4 Protocol Instruction Builder
 *
 * Production-grade instruction builder for Raydium AMM V4 protocol.
 * Supports swap operations with WSOL and USDC pools.
 */

import {
  PublicKey,
  Keypair,
  AccountMeta,
  Instruction,
  SystemProgram,
} from "@solana/web3.js";
import {
  getAssociatedTokenAddressSync,
  createAssociatedTokenAccountInstruction,
  TOKEN_PROGRAM_ID,
  createCloseAccountInstruction,
  NATIVE_MINT,
  createSyncNativeInstruction,
} from "@solana/spl-token";

// ============================================
// Program IDs and Constants
// ============================================

/** Raydium AMM V4 program ID */
export const RAYDIUM_AMM_V4_PROGRAM_ID = new PublicKey(
  "675kPX9MHTjS2zt1qfr1NYHuzeLXfQM9H24wFSUt1Mp8"
);

/** Authority */
export const AUTHORITY = new PublicKey(
  "5Q544fKrFoe6tsEbD7S8EmxGTJYAKtTVhAW5Q5pge4j1"
);

/** WSOL Token Account (mint) */
export const WSOL_TOKEN_ACCOUNT = new PublicKey(
  "So11111111111111111111111111111111111111112"
);

/** USDC Token Account (mint) */
export const USDC_TOKEN_ACCOUNT = new PublicKey(
  "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
);

/** Fee rates */
export const TRADE_FEE_NUMERATOR = 25n;
export const TRADE_FEE_DENOMINATOR = 10000n;
export const SWAP_FEE_NUMERATOR = 25n;
export const SWAP_FEE_DENOMINATOR = 10000n;

// ============================================
// Discriminators
// ============================================

/** Swap base in instruction discriminator (single byte) */
export const SWAP_BASE_IN_DISCRIMINATOR: Buffer = Buffer.from([9]);

/** Swap base out instruction discriminator (single byte) */
export const SWAP_BASE_OUT_DISCRIMINATOR: Buffer = Buffer.from([11]);

// ============================================
// Seeds
// ============================================

export const POOL_SEED = Buffer.from("pool");

// ============================================
// Helper Functions
// ============================================

/**
 * Compute swap amount for AMM V4
 */
export function computeSwapAmount(
  coinReserve: bigint,
  pcReserve: bigint,
  isCoinIn: boolean,
  amountIn: bigint,
  slippageBasisPoints: bigint
): { amountOut: bigint; minAmountOut: bigint } {
  // Apply trade fee (0.25%)
  const amountInAfterFee = amountIn - (amountIn * TRADE_FEE_NUMERATOR) / TRADE_FEE_DENOMINATOR;

  // Calculate output using constant product formula
  let amountOut: bigint;
  if (isCoinIn) {
    // Selling coin for pc: output = (pcReserve * amountIn) / (coinReserve + amountIn)
    const denominator = coinReserve + amountInAfterFee;
    amountOut = (pcReserve * amountInAfterFee) / denominator;
  } else {
    // Selling pc for coin: output = (coinReserve * amountIn) / (pcReserve + amountIn)
    const denominator = pcReserve + amountInAfterFee;
    amountOut = (coinReserve * amountInAfterFee) / denominator;
  }

  // Apply slippage
  const minAmountOut = amountOut - (amountOut * slippageBasisPoints) / 10000n;

  return { amountOut, minAmountOut };
}

/**
 * Create instructions to wrap SOL into WSOL
 */
export function createWsolInstructions(
  payer: PublicKey,
  amount: bigint
): Instruction[] {
  const instructions: Instruction[] = [];
  const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payer, true);

  // Sync native (wrap SOL)
  instructions.push(createSyncNativeInstruction(wsolAta));

  return instructions;
}

/**
 * Create instruction to close WSOL ATA and unwrap to SOL
 */
export function createCloseWsolInstruction(payer: PublicKey): Instruction {
  const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payer, true);
  return createCloseAccountInstruction(wsolAta, payer, payer, [], TOKEN_PROGRAM_ID);
}

/**
 * Create WSOL ATA instruction
 */
export function createWsolAtaInstruction(payer: PublicKey): Instruction {
  const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payer, true);
  return createAssociatedTokenAccountInstruction(
    payer,
    wsolAta,
    payer,
    NATIVE_MINT,
    TOKEN_PROGRAM_ID
  );
}

// ============================================
// Types
// ============================================

export interface RaydiumAmmV4Params {
  amm: PublicKey;
  coinMint: PublicKey;
  pcMint: PublicKey;
  tokenCoin: PublicKey;
  tokenPc: PublicKey;
  coinReserve: bigint;
  pcReserve: bigint;
}

export interface BuildBuyInstructionsParams {
  payer: Keypair | PublicKey;
  outputMint: PublicKey;
  inputAmount: bigint;
  slippageBasisPoints?: bigint;
  fixedOutputAmount?: bigint;
  createInputMintAta?: boolean;
  createOutputMintAta?: boolean;
  closeInputMintAta?: boolean;
  protocolParams: RaydiumAmmV4Params;
}

export interface BuildSellInstructionsParams {
  payer: Keypair | PublicKey;
  inputMint: PublicKey;
  inputAmount: bigint;
  slippageBasisPoints?: bigint;
  fixedOutputAmount?: bigint;
  createOutputMintAta?: boolean;
  closeOutputMintAta?: boolean;
  closeInputMintAta?: boolean;
  protocolParams: RaydiumAmmV4Params;
}

// ============================================
// Instruction Builders
// ============================================

/**
 * Build buy instructions for Raydium AMM V4 protocol
 */
export function buildBuyInstructions(
  params: BuildBuyInstructionsParams
): Instruction[] {
  const {
    payer,
    outputMint,
    inputAmount,
    slippageBasisPoints = 1000n,
    fixedOutputAmount,
    createInputMintAta = true,
    createOutputMintAta = true,
    closeInputMintAta = false,
    protocolParams,
  } = params;

  if (inputAmount === 0n) {
    throw new Error("Amount cannot be zero");
  }

  const payerPubkey = payer instanceof Keypair ? payer.publicKey : payer;
  const instructions: Instruction[] = [];

  const {
    amm,
    coinMint,
    pcMint,
    tokenCoin,
    tokenPc,
    coinReserve,
    pcReserve,
  } = protocolParams;

  // Check pool type
  const isWsol = coinMint.equals(WSOL_TOKEN_ACCOUNT) || pcMint.equals(WSOL_TOKEN_ACCOUNT);
  const isUsdc = coinMint.equals(USDC_TOKEN_ACCOUNT) || pcMint.equals(USDC_TOKEN_ACCOUNT);

  if (!isWsol && !isUsdc) {
    throw new Error("Pool must contain WSOL or USDC");
  }

  // Determine swap direction
  const isBaseIn = coinMint.equals(WSOL_TOKEN_ACCOUNT) || coinMint.equals(USDC_TOKEN_ACCOUNT);

  // Calculate output
  const swapResult = computeSwapAmount(
    coinReserve,
    pcReserve,
    isBaseIn,
    inputAmount,
    slippageBasisPoints
  );
  const minimumAmountOut = fixedOutputAmount || swapResult.minAmountOut;

  // Determine input/output mints
  const inputMint = isWsol ? WSOL_TOKEN_ACCOUNT : USDC_TOKEN_ACCOUNT;

  // Derive user token accounts
  const userSourceTokenAccount = getAssociatedTokenAddressSync(
    inputMint,
    payerPubkey,
    true,
    TOKEN_PROGRAM_ID
  );
  const userDestinationTokenAccount = getAssociatedTokenAddressSync(
    outputMint,
    payerPubkey,
    true,
    TOKEN_PROGRAM_ID
  );

  // Handle WSOL wrapping
  if (createInputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createAssociatedTokenAccountInstruction(
        payerPubkey,
        wsolAta,
        payerPubkey,
        NATIVE_MINT,
        TOKEN_PROGRAM_ID
      )
    );
    instructions.push(createSyncNativeInstruction(wsolAta));
  }

  // Create output mint ATA if needed
  if (createOutputMintAta) {
    instructions.push(
      createAssociatedTokenAccountInstruction(
        payerPubkey,
        userDestinationTokenAccount,
        payerPubkey,
        outputMint,
        TOKEN_PROGRAM_ID
      )
    );
  }

  // Build instruction data (1 byte discriminator + 8 bytes amountIn + 8 bytes minAmountOut)
  const data = Buffer.alloc(17);
  SWAP_BASE_IN_DISCRIMINATOR.copy(data, 0);
  data.writeBigUInt64LE(inputAmount, 1);
  data.writeBigUInt64LE(minimumAmountOut, 9);

  // Build accounts (Raydium AMM V4 has a specific account order)
  const accounts: AccountMeta[] = [
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: amm, isSigner: false, isWritable: true },
    { pubkey: AUTHORITY, isSigner: false, isWritable: false },
    { pubkey: amm, isSigner: false, isWritable: true }, // Amm Open Orders (same as amm for simplicity)
    { pubkey: tokenCoin, isSigner: false, isWritable: true }, // Pool Coin Token Account
    { pubkey: tokenPc, isSigner: false, isWritable: true }, // Pool Pc Token Account
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Program (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Market (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Bids (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Asks (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Event Queue (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Coin Vault Account (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Pc Vault Account (placeholder)
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Vault Signer (placeholder)
    { pubkey: userSourceTokenAccount, isSigner: false, isWritable: true },
    { pubkey: userDestinationTokenAccount, isSigner: false, isWritable: true },
    { pubkey: payerPubkey, isSigner: true, isWritable: true },
  ];

  instructions.push(
    new Instruction({
      keys: accounts,
      programId: RAYDIUM_AMM_V4_PROGRAM_ID,
      data,
    })
  );

  // Close WSOL ATA if requested
  if (closeInputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createCloseAccountInstruction(wsolAta, payerPubkey, payerPubkey, [], TOKEN_PROGRAM_ID)
    );
  }

  return instructions;
}

/**
 * Build sell instructions for Raydium AMM V4 protocol
 */
export function buildSellInstructions(
  params: BuildSellInstructionsParams
): Instruction[] {
  const {
    payer,
    inputMint,
    inputAmount,
    slippageBasisPoints = 1000n,
    fixedOutputAmount,
    createOutputMintAta = true,
    closeOutputMintAta = false,
    closeInputMintAta = false,
    protocolParams,
  } = params;

  if (inputAmount === 0n) {
    throw new Error("Amount cannot be zero");
  }

  const payerPubkey = payer instanceof Keypair ? payer.publicKey : payer;
  const instructions: Instruction[] = [];

  const {
    amm,
    coinMint,
    pcMint,
    tokenCoin,
    tokenPc,
    coinReserve,
    pcReserve,
  } = protocolParams;

  // Check pool type
  const isWsol = coinMint.equals(WSOL_TOKEN_ACCOUNT) || pcMint.equals(WSOL_TOKEN_ACCOUNT);
  const isUsdc = coinMint.equals(USDC_TOKEN_ACCOUNT) || pcMint.equals(USDC_TOKEN_ACCOUNT);

  if (!isWsol && !isUsdc) {
    throw new Error("Pool must contain WSOL or USDC");
  }

  // Determine swap direction (selling token for WSOL/USDC means pc is output)
  const isBaseIn = pcMint.equals(WSOL_TOKEN_ACCOUNT) || pcMint.equals(USDC_TOKEN_ACCOUNT);

  // Calculate output
  const swapResult = computeSwapAmount(
    coinReserve,
    pcReserve,
    isBaseIn,
    inputAmount,
    slippageBasisPoints
  );
  const minimumAmountOut = fixedOutputAmount || swapResult.minAmountOut;

  // Determine output mint
  const outputMint = isWsol ? WSOL_TOKEN_ACCOUNT : USDC_TOKEN_ACCOUNT;

  // Derive user token accounts
  const userSourceTokenAccount = getAssociatedTokenAddressSync(
    inputMint,
    payerPubkey,
    true,
    TOKEN_PROGRAM_ID
  );
  const userDestinationTokenAccount = getAssociatedTokenAddressSync(
    outputMint,
    payerPubkey,
    true,
    TOKEN_PROGRAM_ID
  );

  // Create WSOL ATA for receiving if needed
  if (createOutputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createAssociatedTokenAccountInstruction(
        payerPubkey,
        wsolAta,
        payerPubkey,
        NATIVE_MINT,
        TOKEN_PROGRAM_ID
      )
    );
  }

  // Build instruction data
  const data = Buffer.alloc(17);
  SWAP_BASE_IN_DISCRIMINATOR.copy(data, 0);
  data.writeBigUInt64LE(inputAmount, 1);
  data.writeBigUInt64LE(minimumAmountOut, 9);

  // Build accounts
  const accounts: AccountMeta[] = [
    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    { pubkey: amm, isSigner: false, isWritable: true },
    { pubkey: AUTHORITY, isSigner: false, isWritable: false },
    { pubkey: amm, isSigner: false, isWritable: true }, // Amm Open Orders
    { pubkey: tokenCoin, isSigner: false, isWritable: true }, // Pool Coin Token Account
    { pubkey: tokenPc, isSigner: false, isWritable: true }, // Pool Pc Token Account
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Program
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Market
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Bids
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Asks
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Event Queue
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Coin Vault Account
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Pc Vault Account
    { pubkey: amm, isSigner: false, isWritable: false }, // Serum Vault Signer
    { pubkey: userSourceTokenAccount, isSigner: false, isWritable: true },
    { pubkey: userDestinationTokenAccount, isSigner: false, isWritable: true },
    { pubkey: payerPubkey, isSigner: true, isWritable: true },
  ];

  instructions.push(
    new Instruction({
      keys: accounts,
      programId: RAYDIUM_AMM_V4_PROGRAM_ID,
      data,
    })
  );

  // Close WSOL ATA if requested
  if (closeOutputMintAta && isWsol) {
    const wsolAta = getAssociatedTokenAddressSync(NATIVE_MINT, payerPubkey, true);
    instructions.push(
      createCloseAccountInstruction(wsolAta, payerPubkey, payerPubkey, [], TOKEN_PROGRAM_ID)
    );
  }

  // Close input token ATA if requested
  if (closeInputMintAta) {
    instructions.push(
      createCloseAccountInstruction(
        userSourceTokenAccount,
        payerPubkey,
        payerPubkey,
        [],
        TOKEN_PROGRAM_ID
      )
    );
  }

  return instructions;
}
